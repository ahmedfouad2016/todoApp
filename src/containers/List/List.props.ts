import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {ParamTypes} from 'navigation/ParamTypes';

export interface Props extends NativeStackScreenProps<ParamTypes> {}
