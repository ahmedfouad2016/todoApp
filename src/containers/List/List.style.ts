import {StyleSheet} from 'react-native';
import {colors} from 'utils/colors';

export default StyleSheet.create({
  row: {
    flexDirection: 'row',
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 20,
  },
  icon: {
    tintColor: colors.sky,
  },
  input: {
    width: 300,
    marginRight: 10,
  },
  item: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginHorizontal: 20,
    backgroundColor: colors.lightGray,
    borderRadius: 15,
    padding: 10,
    marginVertical: 5,
  },
  header: {
    marginBottom: 30,
  },
});
