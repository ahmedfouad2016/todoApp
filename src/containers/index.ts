export * from './List';
export * from './ListDetails';
