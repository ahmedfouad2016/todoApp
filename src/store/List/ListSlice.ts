import {createSlice, PayloadAction} from '@reduxjs/toolkit';

export interface todoType {
  id: string;
  description: string;
  completed: boolean;
}

export interface listType {
  id: string;
  title: string;
  todoList: todoType[];
}

interface listState {
  lists: listType[];
}

const initialState: listState = {
  lists: [],
};

const listSlice = createSlice({
  name: 'list',
  initialState,
  reducers: {
    addList(state, action: PayloadAction<string>) {
      state.lists.push({
        id: new Date().getTime().toString(),
        title: action.payload,
        todoList: [],
      });
    },
    editList(state, action: PayloadAction<listType>) {
      const index = state.lists.findIndex(val => val.id === action.payload.id);
      state.lists[index] = {
        ...state.lists[index],
        title: action.payload.title,
      };
    },
    removeList(state, action: PayloadAction<listType>) {
      const index = state.lists.findIndex(val => val.id === action.payload.id);
      state.lists.splice(index, 1);
    },
    addTodo(state, action: PayloadAction<{todo: todoType; list: listType}>) {
      const index = state.lists.findIndex(
        val => val.id === action.payload.list.id,
      );
      state.lists[index].todoList.push({
        ...action.payload.todo,
        id: new Date().getTime().toString(),
      });
    },
    updateTodo(state, action: PayloadAction<{todo: todoType; list: listType}>) {
      const listIndex = state.lists.findIndex(
        val => val.id === action.payload.list.id,
      );
      const todoIndex = state.lists[listIndex].todoList.findIndex(
        val => val.id === action.payload.todo.id,
      );
      state.lists[listIndex].todoList[todoIndex] = {
        ...action.payload.todo,
      };
    },
    deleteTodo(state, action: PayloadAction<{todo: todoType; list: listType}>) {
      const listIndex = state.lists.findIndex(
        val => val.id === action.payload.list.id,
      );
      const todoIndex = state.lists[listIndex].todoList.findIndex(
        val => val.id === action.payload.todo.id,
      );
      state.lists[listIndex].todoList.splice(todoIndex, 1);
    },
  },
});

export const {addList, editList, deleteTodo, removeList, addTodo, updateTodo} =
  listSlice.actions;
export default listSlice.reducer;
