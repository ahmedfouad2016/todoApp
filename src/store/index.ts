import {configureStore, combineReducers} from '@reduxjs/toolkit';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import listReducer from './List/ListSlice';

const rootReducer = combineReducers({
  lists: listReducer,
});

const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage,
  blacklist: ['navigation'],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  devTools: __DEV__,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export const persistedStore = persistStore(store);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
