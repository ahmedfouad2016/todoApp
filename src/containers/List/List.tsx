import React, {useState} from 'react';
import {FlatList, Pressable, View, Text, Alert} from 'react-native';
import {translate} from 'i18n';
import {Icon, TextInput} from 'components';
import {useAppDispatch, useAppSelector} from 'utils/hooks';
import {addList, listType} from 'store/List/ListSlice';
import styles from './List.style';
import {Props} from './List.props';

function List({navigation}: Props) {
  const [listName, setListName] = useState('');
  const dispatch = useAppDispatch();
  const lists = useAppSelector(state => state.lists.lists);

  const navigateToTodo = (list: listType) =>
    navigation.navigate('ListDetails', {list});

  const renderHeader = () => {
    return (
      <View style={styles.row}>
        <TextInput
          value={listName}
          onChangeText={text => setListName(text)}
          inputStyle={styles.input}
          placeholder={translate('list.placeholder')}
        />
        <Pressable
          testID="add-list-btn"
          onPress={() => {
            if (listName) {
              dispatch(addList(listName));
              setListName('');
            } else {
              Alert.alert(
                translate('common.warning'),
                translate('list.nameRequired'),
              );
            }
          }}>
          <Icon icon="add" style={styles.icon} size={45} />
        </Pressable>
      </View>
    );
  };
  return (
    <FlatList
      data={lists}
      renderItem={({item}) => (
        <Pressable style={styles.item} onPress={() => navigateToTodo(item)}>
          <Text>{item.title}</Text>
          <Icon icon="arrowRight" size={40} style={styles.icon} />
        </Pressable>
      )}
      ListHeaderComponentStyle={styles.header}
      ListHeaderComponent={renderHeader()}
    />
  );
}

export {List};
