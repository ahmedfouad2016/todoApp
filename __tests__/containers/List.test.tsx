import 'react-native';
import React from 'react';
import configureStore from 'redux-mock-store';
import {act, fireEvent, render} from '@testing-library/react-native';
import {List} from 'containers/List';
import {Provider} from 'react-redux';
import {addList} from 'store/List/ListSlice';
import {store as ReduxStore} from 'store';

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...props,
});

describe('Todo Lists Screen', () => {
  const mockStore = configureStore([]);
  let props: any;

  it('should List render', () => {
    const initialState = {
      lists: {
        lists: [],
      },
    };
    const store = mockStore(initialState);
    props = createTestProps({});
    const wrapper = render(
      <Provider store={store}>
        <List {...props} />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should List to have 3 todo lists', () => {
    const title = 'My list';
    const initialState = {
      lists: {
        lists: [
          {id: '1', title: title, todoList: []},
          {id: '2', title: title, todoList: []},
          {id: '3', title: title, todoList: []},
        ],
      },
    };
    const store = mockStore(initialState);
    props = createTestProps({});
    const {queryAllByText} = render(
      <Provider store={store}>
        <List {...props} />
      </Provider>,
    );
    expect(queryAllByText(title).length).toEqual(3);
  });

  it('should navigate to todo screen on press the list', () => {
    const initialState = {
      lists: {
        lists: [{id: '1', title: 'my list', todoList: []}],
      },
    };
    const store = mockStore(initialState);
    props = createTestProps({});
    const {getByText} = render(
      <Provider store={store}>
        <List {...props} />
      </Provider>,
    );
    const ele = getByText('my list').parent || getByText('my list');
    fireEvent.press(ele);
    expect(props.navigation.navigate).toHaveBeenCalled();
  });

  it('should addList action fire after press the add btn', () => {
    const initialState = {
      lists: {
        lists: [],
      },
    };
    const store = mockStore(initialState);
    props = createTestProps({});
    const {getByTestId} = render(
      <Provider store={store}>
        <List {...props} />
      </Provider>,
    );
    fireEvent.changeText(getByTestId('text-input'), 'my list');
    fireEvent.press(getByTestId('add-list-btn'));
    const actions = store.getActions();
    const expectedAction = {type: 'list/addList', payload: 'my list'};
    expect(actions).toContainEqual(expectedAction);
  });

  it('should list add to todo lists after dispatch addList action', () => {
    props = createTestProps({});
    const {queryAllByText} = render(
      <Provider store={ReduxStore}>
        <List {...props} />
      </Provider>,
    );
    const payload = 'my list';
    act(() => {
      ReduxStore.dispatch(addList(payload));
      console.log(ReduxStore.getState().lists.lists);
    });
    expect(queryAllByText(payload).length).toEqual(1);
  });
});
