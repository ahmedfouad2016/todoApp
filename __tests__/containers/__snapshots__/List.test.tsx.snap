// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Lists Screen should List render 1`] = `
<RCTScrollView
  ListHeaderComponent={
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 20,
        }
      }
    >
      <TextInput
        inputStyle={
          Object {
            "marginRight": 10,
            "width": 300,
          }
        }
        onChangeText={[Function]}
        placeholder="list.placeholder.test"
        value=""
      />
      <Memo(Pressable)
        onPress={[Function]}
        testID="add-list-btn"
      >
        <Icon
          icon="add"
          size={45}
          style={
            Object {
              "tintColor": "#4498D1",
            }
          }
        />
      </Memo(Pressable)>
    </View>
  }
  ListHeaderComponentStyle={
    Object {
      "marginBottom": 30,
    }
  }
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "marginBottom": 30,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 20,
          }
        }
      >
        <TextInput
          onChangeText={[Function]}
          placeholder="list.placeholder.test"
          style={
            Object {
              "borderWidth": 1,
              "height": 40,
              "marginRight": 10,
              "padding": 10,
              "width": 300,
            }
          }
          testID="text-input"
          value=""
        />
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="add-list-btn"
        >
          <View>
            <RCTImageView
              internal_analyticTag={null}
              resizeMode="contain"
              source={
                Array [
                  Object {
                    "testUri": "../../../src/assets/icons/add.png",
                  },
                ]
              }
              style={
                Object {
                  "height": 45,
                  "overflow": "hidden",
                  "resizeMode": "contain",
                  "tintColor": "#4498D1",
                  "width": 45,
                }
              }
              testID="icon-image"
              tintColor="#4498D1"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
